### PR TITLE
DAOS-3749 test: Verify aggregation throttling

### PR DESCRIPTION
**Pull request by Puspalogan**
- Created aggregation/aggregation_throttling test to
    verify the aggregation throttling feature

Signed-off-by: Logan Sundaram <logan.sundaram@intel.com>

### CODE REVIEW

File: src/tests/ftest/aggregation/aggregation_throttling.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+                                10.0) # 10% perf difference
+
+        self.verify_performance(metric_before_aggregate,
+                                metric_after_aggregate,
+                                1, # read_perf
+                                10.0)
+
+    def verify_performance(self, before_metric, after_metric, read_write_idx,
+                           expected_perf_diff):
+        """Verify the after_metric read/write performances are within
```
[the remark this answers is not in the document]
Comment (rpadma2) [reply src/tests/ftest/aggregation/aggregation_throttling.py:100]: Please fix the trailing white space issue.

File: src/tests/ftest/aggregation/aggregation_throttling.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        self.verify_performance(metric_before_aggregate,
+                                metric_after_aggregate,
+                                1, # read_perf
+                                10.0)
+
+    def verify_performance(self, before_metric, after_metric, read_write_idx,
+                           expected_perf_diff):
+        """Verify the after_metric read/write performances are within 
+           +/- expected_perf_diff % of before_metric_performance.
+
```
[the remark this answers is not in the document]
Comment (rpadma2) [reply src/tests/ftest/aggregation/aggregation_throttling.py:102]: Ditto.

**Comment by rpadma2** [comment]
Did you try aggregation difference on different object class (eg: RP_GP1, RPGP2,etc)? What about multiple pools in a cluster and aggregation performance? If not, this is some of the enhancements you can consider for future testing.

File: src/tests/ftest/aggregation/aggregation_throttling.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Puspalogan) [reply src/tests/ftest/aggregation/aggregation_throttling.py]: Done

File: src/tests/ftest/aggregation/aggregation_throttling.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Puspalogan) [reply src/tests/ftest/aggregation/aggregation_throttling.py]: Done

File: src/tests/ftest/aggregation/aggregation_throttling.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        metric_after_aggregate = IorCommand.get_ior_metrics(out)
+
+        # When DAOS-5057 is fixed, adjust the percentage. For now, 
+        # keep it at 30 %
+        expected_perf_diff = 30.0
+
+        self.verify_performance(metric_before_aggregate,
+                                metric_after_aggregate,
+                                0, # write_perf
+                                expected_perf_diff) # 10% perf difference
```
Comment (saurabhtandan) [inline src/tests/ftest/aggregation/aggregation_throttling.py:97]: For Line 91 above, the perf diff mentioned is 30%, so may be for the comment in this line you might want to change it to 30% from 10%.

I did saw the ticket mentions 10% perf diff and later there is a comment by you that write perf is 30% diff. Is this acceptable number as per developer or should there be a parallel ticket to handle this issue?

File: src/tests/ftest/aggregation/aggregation_throttling.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+        Args:
+           before_metric (IorMetrics): ior metrics before aggregation
+           after_metric (IorMetrics): ior metrics in concurrent with aggregation
+           read_write_idx (int): Read (1) or Write (0) index in the IorMetrics
+           expected_perf_diff (float): Expected performance difference between
+                                       before_metric and after_metric.
+        """
+
+        self.log.info("\n\n{} Performance".format(
```
[the remark this answers is not in the document]
Comment (saurabhtandan) [reply src/tests/ftest/aggregation/aggregation_throttling.py:117]: May be you can resolve this noisy lint as follows:
self.log.info("\n\n%s Performance", "Read" if read_write_idx else "Write")

File: src/tests/ftest/aggregation/aggregation_throttling.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+            "Read" if read_write_idx else "Write"))
+
+        max_mib = int(IorMetrics.Max_MiB)
+        min_mib = int(IorMetrics.Min_MiB)
+        mean_mib = int(IorMetrics.Mean_MiB)
+
+        max_prev = float(before_metric[read_write_idx][max_mib])
+        max_curr = float(after_metric[read_write_idx][max_mib])
+        self.log.info("max_prev = {}, max_curr = {}".format(
```
[the remark this answers is not in the document]
Comment (saurabhtandan) [reply src/tests/ftest/aggregation/aggregation_throttling.py:126]: Same as above

File: src/tests/ftest/aggregation/aggregation_throttling.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            max_prev, max_curr))
+
+        min_prev = float(before_metric[read_write_idx][min_mib])
+        min_curr = float(after_metric[read_write_idx][min_mib])
+        self.log.info("min_prev = {}, min_curr = {}".format(
```
[the remark this answers is not in the document]
Comment (saurabhtandan) [reply src/tests/ftest/aggregation/aggregation_throttling.py:131]: Same as above

File: src/tests/ftest/aggregation/aggregation_throttling.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+            min_prev, min_curr))
+
+        mean_prev = float(before_metric[read_write_idx][mean_mib])
+        mean_curr = float(after_metric[read_write_idx][mean_mib])
+        self.log.info("mean_prev = {}, mean_curr = {}".format(
```
[the remark this answers is not in the document]
Comment (saurabhtandan) [reply src/tests/ftest/aggregation/aggregation_throttling.py:136]: Same as above

File: src/tests/ftest/aggregation/aggregation_throttling.py (hunk 3 above)
Comment (Puspalogan) [reply src/tests/ftest/aggregation/aggregation_throttling.py:97]: JIRA # DAOS-5057 is there to look into the performance issue. But to get this ticket to CI, 30% diff is acceptable for now per Johann's comment. Will need to revisit based on DAOS-5057 resolution.

**Comment by Puspalogan** [comment]
Func HW Small issue due to - DAOS-5321
Func HW Large issue due to - DAOS-5383

File: src/tests/ftest/aggregation/aggregation_throttling.py [hunk 8]
```diff
@@ -35,13 +36,14 @@ class DaosAggregationThrottling(IorTestBase):
     :avocado: recursive
     """
 
+    @skipForTicket("DAOS-")
```
Comment (shimizukko) [inline src/tests/ftest/aggregation/aggregation_throttling.py:39]: Is there going to be a ticket for this?

Comment (Puspalogan) [reply src/tests/ftest/aggregation/aggregation_throttling.py]: Was going to skip it but now adjusted the performance until DAOS-5057 ticket is resolved. Removed this decorator.

**Comment by Puspalogan** [comment]
Functional HW Large failure is due to DAOS-5388

**Comment by Puspalogan** [comment]
> LGTM
> But I would recommend as this test is really long (21 Mints), we should have this test in weekly run instead of PR. This will increase the CI testing time and queue considerably.

Sure, Changed it to run in weekly.

**Comment by Puspalogan** [comment]
HW Func Large failure is not related to this PR. There's a ticket DAOS-5383 exist for that.

**Pull request merged**